File: reverse/tools/ghidra_decompile_all.py
# Ghidra headless post-script: decompile every function to C-like text
# @category NDM
# @runtime Jython

from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor
import os

out_root = os.environ.get(
    "NDM_GHIDRA_OUT",
    "/Users/gaoyuan/NDM/reverse/dumps/full_decompile/ghidra_c",
)
if not os.path.isdir(out_root):
    os.makedirs(out_root)

program = currentProgram
listing = program.getListing()
fm = program.getFunctionManager()
monitor = ConsoleTaskMonitor()

ifc = DecompInterface()
ifc.openProgram(program)

index_path = os.path.join(out_root, "INDEX.tsv")
index_f = open(index_path, "w")
index_f.write("address\tname\tfile\tsize\tsuccess\n")

count = 0
ok = 0
fail = 0

funcs = fm.getFunctions(True)
for func in funcs:
    count += 1
    name = func.getName()
    entry = func.getEntryPoint()
    addr_s = "0x%X" % entry.getOffset()
    safe = "".join(c if (c.isalnum() or c in "._-") else "_" for c in name)[:100]
    fname = "%s__%s.c" % (safe, addr_s)
    path = os.path.join(out_root, fname)

    results = ifc.decompileFunction(func, 60, monitor)
    success = results is not None and results.decompileCompleted()
    body = ""
    if success:
        try:
            body = results.getDecompiledFunction().getC()
            ok += 1
        except Exception as e:
            body = "// decompile completed but getC failed: %s\n" % e
            success = False
            fail += 1
    else:
        fail += 1
        err = ""
        try:
            err = str(results.getErrorMessage())
        except Exception:
            err = "unknown"
        body = "// DECOMPILE FAILED: %s\n" % err
        # still dump disassembly fallback via listing
        try:
            insts = listing.getInstructions(func.getBody(), True)
            lines = ["// fallback disassembly for %s\n" % name]
            while insts.hasNext() and len(lines) < 5000:
                ins = insts.next()
                lines.append("%s  %s\n" % (ins.getAddress(), ins.toString()))
            body += "".join(lines)
        except Exception as e2:
            body += "// disasm fallback failed: %s\n" % e2

    with open(path, "w") as f:
        f.write("/* function %s @ %s */\n" % (name, addr_s))
        f.write(body if body else "// empty\n")

    index_f.write("%s\t%s\t%s\t%d\t%s\n" % (
        addr_s, name, fname, func.getBody().getNumAddresses(), "1" if success else "0"
    ))
    if count % 100 == 0:
        print("progress %d ok=%d fail=%d" % (count, ok, fail))

index_f.close()
summary = "total=%d ok=%d fail=%d out=%s\n" % (count, ok, fail, out_root)
with open(os.path.join(out_root, "SUMMARY.txt"), "w") as f:
    f.write(summary)
print(summary)

File: reverse/tools/GhidraDecompileAll.java
//Decompile every function in the current program to C-like text files.
//@category NDM
//@menupath Tools.NDM.DecompileAll

import java.io.File;
import java.io.FileWriter;
import java.io.PrintWriter;

import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.app.script.GhidraScript;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionIterator;
import ghidra.program.model.listing.Instruction;
import ghidra.program.model.listing.InstructionIterator;
import ghidra.program.model.listing.Listing;
import ghidra.util.task.ConsoleTaskMonitor;

public class GhidraDecompileAll extends GhidraScript {

	@Override
	protected void run() throws Exception {
		String outRoot = System.getenv("NDM_GHIDRA_OUT");
		if (outRoot == null || outRoot.isEmpty()) {
			outRoot = "/Users/gaoyuan/NDM/reverse/dumps/full_decompile/ghidra_c";
		}
		File outDir = new File(outRoot);
		if (!outDir.exists() && !outDir.mkdirs()) {
			printerr("cannot create " + outRoot);
			return;
		}

		DecompInterface ifc = new DecompInterface();
		ifc.openProgram(currentProgram);
		ConsoleTaskMonitor monitor = new ConsoleTaskMonitor();
		Listing listing = currentProgram.getListing();

		File indexFile = new File(outDir, "INDEX.tsv");
		PrintWriter index = new PrintWriter(new FileWriter(indexFile));
		index.println("address\tname\tfile\tbytes\tsuccess");

		int count = 0;
		int ok = 0;
		int fail = 0;

		FunctionIterator funcs = currentProgram.getFunctionManager().getFunctions(true);
		while (funcs.hasNext() && !monitor.isCancelled()) {
			Function func = funcs.next();
			count++;
			String name = func.getName();
			String addrS = String.format("0x%X", func.getEntryPoint().getOffset());
			String safe = name.replaceAll("[^A-Za-z0-9._-]", "_");
			if (safe.length() > 100) {
				safe = safe.substring(0, 100);
			}
			String fname = safe + "__" + addrS + ".c";
			File out = new File(outDir, fname);

			DecompileResults results = ifc.decompileFunction(func, 60, monitor);
			boolean success = results != null && results.decompileCompleted();
			String body;
			if (success && results.getDecompiledFunction() != null) {
				body = results.getDecompiledFunction().getC();
				ok++;
			}
			else {
				fail++;
				String err = results == null ? "null" : String.valueOf(results.getErrorMessage());
				StringBuilder sb = new StringBuilder();
				sb.append("// DECOMPILE FAILED: ").append(err).append("\n");
				sb.append("// fallback disassembly\n");
				try {
					InstructionIterator insts = listing.getInstructions(func.getBody(), true);
					int n = 0;
					while (insts.hasNext() && n < 5000) {
						Instruction ins = insts.next();
						sb.append(ins.getAddress()).append("  ").append(ins.toString()).append("\n");
						n++;
					}
				}
				catch (Exception e) {
					sb.append("// disasm fallback failed: ").append(e).append("\n");
				}
				body = sb.toString();
			}

			PrintWriter pw = new PrintWriter(new FileWriter(out));
			pw.println("/* function " + name + " @ " + addrS + " */");
			pw.print(body == null ? "// empty\n" : body);
			pw.close();

			long nbytes = func.getBody().getNumAddresses();
			index.println(addrS + "\t" + name + "\t" + fname + "\t" + nbytes + "\t" + (success ? "1" : "0"));

			if (count % 100 == 0) {
				println("progress " + count + " ok=" + ok + " fail=" + fail);
			}
		}

		index.close();
		PrintWriter sum = new PrintWriter(new FileWriter(new File(outDir, "SUMMARY.txt")));
		String summary = "total=" + count + " ok=" + ok + " fail=" + fail + " out=" + outRoot;
		sum.println(summary);
		sum.close();
		println(summary);
	}
}
